File: auto_outsrc/parser/BGW/forAyoBGW.py
import builtInFuncs
from builtInFuncs import *
from charm.toolbox.pairinggroup import PairingGroup,ZR,G1,G2,GT,pair
group = None

secparam = 80


def setup(n):
    glG1 = {}
    glG2 = {}

    gG1 = group.random(G1)
    gG2 = group.random(G2)
    alpha = group.random(ZR)
    index = ((2 * n) + 1)
    for i in range(1, index):
        glG1[i] = (gG1 ** (alpha ** i))
        glG2[i] = (gG2 ** (alpha ** i))
    gamma = group.random(ZR)
    v = (gG2 ** gamma)
    pk = [gG1, gG2, glG1, glG2, v]
    msk = [gamma]
    output = (pk, msk, n)
    return output

def keygen(pk, msk, n):
    sBlinded = {}
    s = {}

    bf0 = group.random(ZR)
    gG1, gG2, glG1, glG2, v = pk
    [gamma] = msk
    for i in range(1, n+1):
        s[i] = (glG1[i] ** gamma)
    for y in s:
        sBlinded[y] = (s[y] ** (1 / bf0))
    skBlinded = [sBlinded]
    output = (bf0, skBlinded)
    return output

def encrypt(S, pk, n):
    gG1, gG2, glG1, glG2, v = pk
    t = group.random(ZR)
    K = (pair(glG1[n], glG2[1]) ** t)
    dotProd1 = group.init(G2)
    for j in S:
        dotProd1 = (dotProd1 * glG2[n+1-j])
    Hdr2 = ((v * dotProd1) ** t)
    Hdr1 = (gG2 ** t)
    Hdr = [Hdr1, Hdr2]
    ct = [Hdr, K]
    output = ct
    return output

def transform(S, i, n, Hdr, pk, skBlinded):
    transformOutputList = {}
    transformOutputListForLoop = {}

    Hdr1, Hdr2 = Hdr
    gG1, gG2, glG1, glG2, v = pk
    [sBlinded] = skBlinded
    transformOutputList[0] = pair(glG1[i], Hdr2)
    numerator = transformOutputList[0]
    transformOutputList[1] = group.init(G1)
    dotProd2 = transformOutputList[1]
    lenS = len(S)
    for k in range(0, lenS):
        pass
        j = S[k]
        if ( ( (j) != (i) ) ):
            pass
            FLrepVar2 = (10 + (5 * k))
            transformOutputListForLoop[FLrepVar2] = (dotProd2 * glG1[n+1-j+i])
            dotProd2 = transformOutputListForLoop[FLrepVar2]
    transformOutputList[2] = pair(sBlinded[i], Hdr1)
    transformOutputList[3] = pair(dotProd2, Hdr1)
    output = (transformOutputList)
    return output

def decout(S, i, n, Hdr, pk, transformOutputList, bf0):
    Hdr1, Hdr2 = Hdr
    gG1, gG2, glG1, glG2, v = pk
    numerator = transformOutputList[0]
    dotProd2 = transformOutputList[1]
    denominator = ((transformOutputList[2] ** bf0) * transformOutputList[3])
    K = (numerator * (denominator ** -1))
    output = K
    return output

def main():
    global group
    #group = PairingGroup(secparam)
    group = PairingGroup('SS512')

    (pk, msk, n) = setup(15)
    (blindingFactor0Blinded, skComplete) = keygen(pk, msk, n)
    S = [1, 3, 5, 12, 14]
    (Hdr, K) = encrypt(S, pk, n)
    print("K:  ", K)
    i = 1
    transformOutputList = transform(S, i, n, Hdr, pk, skComplete)
    Krecovered = decout(S, i, n, Hdr, pk, transformOutputList, blindingFactor0Blinded)
    print("Recovered K = ", Krecovered)
    if (K == Krecovered):
        print("Successful")
    else:
        print("Failed")

if __name__ == '__main__':
    main()

File: auto_outsrc/parser/BGW/forAyoBGW.cpp
#include "BGW/forAyoBGW.h"

void Bgw05::setup(int n, CharmList & pk, CharmList & msk)
{
    G1 gG1;
    G2 gG2;
    ZR alpha;
    int index = 0;
    CharmListG1 glG1;
    CharmListG2 glG2;
    ZR gamma;
    G1 v;
    gG1 = group.random(G1_t);
    gG2 = group.random(G2_t);
    alpha = group.random(ZR_t);
    index = group.add((2 * n), 1);
    for (int i = 1; i < index; i++)
    {
        glG1[i] = group.exp(gG1, group.exp(alpha, i));
        glG2[i] = group.exp(gG2, group.exp(alpha, i));
    }
    gamma = group.random(ZR_t);
    v = group.exp(gG1, gamma);
    pk.insert(0, gG1);
    pk.insert(1, gG2);
    pk.insert(2, glG1);
    pk.insert(3, glG2);
    pk.insert(4, v);
    msk.insert(0, gamma);
    return;
}

void Bgw05::keygen(CharmList & pk, CharmList & msk, int n, ZR & bf0, CharmList & skBlinded)
{
    G1 gG1;
    G2 gG2;
    CharmListG1 glG1;
    CharmListG2 glG2;
    G1 v;
    ZR gamma;
    CharmListG2 s;
    CharmListG2 sBlinded;
    bf0 = group.random(ZR_t);
    
    gG1 = pk[0].getG1();
    gG2 = pk[1].getG2();
    glG1 = pk[2].getListG1();
    glG2 = pk[3].getListG2();
    v = pk[4].getG1();
    
    gamma = msk[0].getZR();
    for (int i = 1; i < n+1; i++)
    {
        s[i] = group.exp(glG2[i], gamma);
    }
    CharmListInt s_keys = s.keys();
    int s_len = s_keys.length();
    for (int y_var = 0; y_var < s_len; y_var++)
    {
        int y = s_keys[y_var];
        sBlinded[y] = group.exp(s[y], group.div(1, bf0));
    }
    skBlinded.insert(0, sBlinded);
    return;
}

void Bgw05::encrypt(CharmListInt & S, CharmList & pk, int n, CharmList & ct)
{
    G1 gG1;
    G2 gG2;
    CharmListG1 glG1;
    CharmListG2 glG2;
    G1 v;
    ZR t;
    GT K = group.init(GT_t);
    G1 dotProd1 = group.init(G1_t, 1);
    G1 Hdr2;
    G1 Hdr1;
    CharmList Hdr;
    
    gG1 = pk[0].getG1();
    gG2 = pk[1].getG2();
    glG1 = pk[2].getListG1();
    glG2 = pk[3].getListG2();
    v = pk[4].getG1();
    t = group.random(ZR_t);
    K = group.exp(group.pair(glG1[n], glG2[1]), t);
    group.init(dotProd1, 1);
    CharmListInt S_keys = S.keys();
    int S_len = S_keys.length();
    for (int j_var = 0; j_var < S_len; j_var++)
    {
        int j = S_keys[j_var];
        dotProd1 = group.mul(dotProd1, glG1[n+1-j]);
    }
    Hdr2 = group.exp(group.mul(v, dotProd1), t);
    Hdr1 = group.exp(gG1, t);
    Hdr.insert(0, Hdr1);
    Hdr.insert(1, Hdr2);
    ct.insert(0, Hdr);
    ct.insert(1, K);
    return;
}

void Bgw05::transform(CharmListInt & S, int i, int n, CharmList & Hdr, CharmList & pk, CharmList & skBlinded, CharmList & transformOutputList)
{
    G1 Hdr1;
    G1 Hdr2;
    G1 gG1;
    G2 gG2;
    CharmListG1 glG1;
    CharmListG2 glG2;
    G1 v;
    CharmListG2 sBlinded;
    GT numerator = group.init(GT_t);
    G2 dotProd2 = group.init(G2_t, 1);
    int lenS = 0;
    int j = 0;
    int FLrepVar2 = 0;
    CharmList transformOutputListForLoop;
    
    Hdr1 = Hdr[0].getG1();
    Hdr2 = Hdr[1].getG1();
    
    gG1 = pk[0].getG1();
    gG2 = pk[1].getG2();
    glG1 = pk[2].getListG1();
    glG2 = pk[3].getListG2();
    v = pk[4].getG1();
    
    sBlinded = skBlinded[0].getListG2();
    transformOutputList.insert(0, group.pair(glG2[i], Hdr2));
    numerator = transformOutputList[0].getGT();
    transformOutputList.insert(1, group.init(G2_t));
    dotProd2 = transformOutputList[1].getG2();
    lenS = S.length();
    for (int k = 0; k < lenS; k++)
    {

        j = S[k];
        if ( ( (j) != (i) ) )
        {

            FLrepVar2 = group.add(10, (5 * k));
            transformOutputListForLoop.insert(FLrepVar2, group.mul(dotProd2, glG2[n+1-j+i]));
            dotProd2 = transformOutputListForLoop[FLrepVar2].getG2().getG2();
        }
    }
    transformOutputList.insert(2, group.pair(Hdr1, sBlinded[i]));
    transformOutputList.insert(3, group.pair(Hdr1, dotProd2));
    return;
}

void Bgw05::decout(CharmListInt & S, int i, int n, CharmList & Hdr, CharmList & pk, CharmList & transformOutputList, ZR & bf0, GT & K)
{
    G1 Hdr1;
    G1 Hdr2;
    G1 gG1;
    G2 gG2;
    CharmListG1 glG1;
    CharmListG2 glG2;
    G1 v;
    GT numerator = group.init(GT_t);
    G2 dotProd2 = group.init(G2_t, 1);
    GT denominator = group.init(GT_t);
    
    Hdr1 = Hdr[0].getG1();
    Hdr2 = Hdr[1].getG1();
    
    gG1 = pk[0].getG1();
    gG2 = pk[1].getG2();
    glG1 = pk[2].getListG1();
    glG2 = pk[3].getListG2();
    v = pk[4].getG1();
    numerator = transformOutputList[0].getGT();
    dotProd2 = transformOutputList[1].getG2();
    denominator = group.mul(group.exp(transformOutputList[2].getGT(), bf0), transformOutputList[3].getGT());
    K = group.mul(numerator, group.exp(denominator, -1));
    return;
}

